File: app/src/main/java/home/oleg/placesnearme/presentation/feature/map/viewmodel/UserLocationViewModel.java
package home.oleg.placesnearme.presentation.feature.map.viewmodel;

import home.oleg.placenearme.interactors.GetUserLocationInteractor;
import home.oleg.placesnearme.presentation.base.BaseViewModel;
import home.oleg.placesnearme.presentation.feature.map.view.UserLocationView;
import home.oleg.placesnearme.presentation.view_action.ShowUserLocationAction;
import home.oleg.placesnearme.presentation.view_action.ViewActions;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by Oleg Sheliakin on 17.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public class UserLocationViewModel extends BaseViewModel<UserLocationView> {

    private final GetUserLocationInteractor userLocationInteractor;

    public UserLocationViewModel(GetUserLocationInteractor userLocationInteractor) {
        this.userLocationInteractor = userLocationInteractor;
    }

    public void getUserLocation() {
        addToDisposables(userLocationInteractor.getUserLocation()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(location -> {
                            setAction(ShowUserLocationAction.create(location));
                        },
                        throwable -> {
                            setAction(ViewActions.error(throwable));
                        })
        );
    }

}

File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/presentation/viewmodel/UserLocationViewModel.kt
package home.oleg.placesnearme.feature_map.presentation.viewmodel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import home.oleg.placenearme.models.UserLocation
import home.oleg.placesnearme.core_presentation.delegate.disposableDelegate
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers

/**
 * Created by Oleg Sheliakin on 17.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
class UserLocationViewModel(private val getLocation: () -> Single<UserLocation>) : ViewModel() {

    private var disposable: Disposable? by disposableDelegate()
    private val stateInternal = MutableLiveData<UserLocation>()

    val location = stateInternal

    fun requestUserLocation() {
        disposable = getLocation()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(
                        onSuccess = { stateInternal.setValue(it) },
                        onError = Throwable::printStackTrace)
    }
}
